micromasters_import/management/commands/queries/008_import_combined_final_grade.sql

```diff
@@ -0,0 +1,106 @@
+----This script is to migrate the grades for course run and exam from MicroMaster:
+----  It computes score from FinalGrade and ProctoredExamGrade using formula - course run grade * 0.4 + exam grade * 0.6
+----  grades need to exist in both FinalGrade (course run) and ProctoredExamGrade (exam)
+----  For 'passed' field, true if both have passed= true, otherwise false.
+
+WITH courses_have_exam AS (
+    SELECT
+        DISTINCT mm_examrun.course_id
+    FROM micromasters.exams_examrun AS mm_examrun
+    JOIN micromasters.courses_course AS mm_course
+        ON mm_examrun.course_id = mm_course.id
+),
+ --- this is to compute the best grade learner gets for each course run regardless if passed or not
+best_courserun_grades AS (
+    SELECT
+        mm_grade.user_id,
+        mm_grade.course_run_id,
+        MAX(mm_grade.grade) AS max_grade
+    FROM micromasters.grades_finalgrade AS mm_grade
+    JOIN micromasters.courses_courserun AS mm_courserun
+        ON mm_grade.course_run_id = mm_courserun.id
+    JOIN courses_have_exam
+        ON mm_courserun.course_id = courses_have_exam.course_id
+    WHERE mm_grade.status = 'complete'
+    GROUP BY mm_grade.user_id, mm_grade.course_run_id
+),
+--- this is to compute the best exam grades learner get for every course run in the exam they signed up for
+best_exam_grades AS (
+    SELECT
+        mm_exam_grade.user_id,
+        mm_exam_grade.course_id,
+        mm_courserun.id AS course_run_id,
+        MAX(mm_exam_grade.score) AS max_score
+    FROM micromasters.grades_proctoredexamgrade AS mm_exam_grade
+    JOIN micromasters.courses_courserun AS mm_courserun
+       ON mm_exam_grade.course_id = mm_courserun.course_id
+    JOIN courses_have_exam
+       ON mm_exam_grade.course_id = courses_have_exam.course_id
+    GROUP BY mm_exam_grade.user_id, mm_courserun.id, mm_exam_grade.course_id
+),
+combined_final_grades AS (
+    SELECT
+        best_grade.user_id,
+        best_grade.course_run_id,
+        ROUND(CAST(best_grade.max_grade * 100 * 0.4 + best_exam.max_score * 0.6 AS numeric), 1) AS calculated_grade,
```

Comment:
Is it true that we don't have a materialized combined_final_grade in the micromasters database? I feel like someone told me that, but it always surprises me.

Reply:
There is a grade field in CombinedFinalGrade in micromasters, but it's at the course level, not the course run level.

Reply:
Did you consider migrating that calculated grade, instead of recalculating? I am pretty sure you have the correct formula here, but I feel like there's always some chance that the data in MM is different from this calculation.

Reply:
@pdpinch I hope maybe this helps clarify what we're trying to do here so you can confirm this is a valid approach. 

To start, there's a difference in how MicroMasters and MITx Online represent grade data in the database because MicroMasters was historically tailored to how DEDP operated these courses and MITx Online departs from that in a way that doesn't map one-to-one:

#### MicroMasters

- Learners can have multiple grades for a course, which are stored in `grades_finalgrade`. These are mapped to a course run.
- Learners can take one or more exams, separately from when they take the course itself. These are stored in `grades_proctoredexamgrade`.
- There is only **one** combined grade tracked in the database, at the course level. This grade is recalculated every time a record in `grades_finalgrade` or `grades_proctoredexamgrade` is created or updated. This is calculated as the weighted combination of the best course run and exam grades.

#### MITx Online

- Courses and exams have been combined on the openedx side and as a result the grades coming back from openedx are the combined and weighted grade.

Additionally, it was my understanding that MITx Online would be handling the grades in the following manner:
- There will be no computation done on the MITx Online side in terms of the "best grade of runs and exams" logic that MicroMasters had since the grades are opaque to us.
- If the course team wants to support some kind of "forwarding of the best grade to future runs" to support something similar to the MicroMasters behavior, that would be their responsibility to manage through openedx functionality and not a feature of MITx Online itself.
- As a result, when we eventually generate program certificates, we'll simply take the best course run grade.

Assuming all the above holds true, we'd been assuming we wanted to made the legacy grades into MITx Online while keeping to MicroMasters' grade calculation logic:

- For each course run `FinalGrade` in Micromasters, create a corresponding `CourseRunGrade` in MITx Online, weighted using the best exam grade.
- This preserves MicroMasters' legacy logic while making it compatible with MITx Online's data structure and also for any future exam attempts a learner might take because MicroMasters' `max(final_grades) * 0.4 + max(exam_grades) * 0.6` is functionally the same as just taking the `max(course_run_grades)` of legacy run grades that have been migrated as `course_run_grade * 0.4 + max(exam_grades) * 0.6`.
- This keeps the query simple because it doesn't need to do a subquery to also determine the best course run grade. Since we'll just end up using the best course run grade anyway on MITx Online, the lower grades get shadowed anyway so there's no harm in bringing them over.